### PR TITLE
etcdmain: add '/metrics' HTTP/1 path to grpc-proxy

### PR DESCRIPTION
**Pull request by gyuho**
Broken. WIP.

Just testing to see what happens in grpcproxy tests.

### CODE REVIEW

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 This works now. Please review.

Fix https://github.com/coreos/etcd/issues/6854.

File: etcdmain/grpc_proxy.go [hunk 1]
```diff
@@ -74,6 +79,16 @@ func startGRPCProxy(cmd *cobra.Command, args []string) {
 		fmt.Fprintln(os.Stderr, err)
 		os.Exit(1)
 	}
+	if l, err = transport.NewKeepAliveListener(l, "tcp", nil); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+	plog.Println("listening for grpc-proxy client requests on", grpcProxyListenAddr)
```
Comment (heyitsanthony) [inline etcdmain/grpc_proxy.go]: `Infof`?

File: etcdmain/grpc_proxy.go [hunk 2]
```diff
@@ -94,15 +109,34 @@ func startGRPCProxy(cmd *cobra.Command, args []string) {
...
 	pb.RegisterKVServer(server, kvp)
 	pb.RegisterWatchServer(server, watchp)
 	pb.RegisterClusterServer(server, clusterp)
 	pb.RegisterLeaseServer(server, leasep)
 	pb.RegisterMaintenanceServer(server, mainp)
 	pb.RegisterAuthServer(server, authp)
+	grpc_prometheus.Register(server)
+
+	grpcl := m.Match(cmux.HTTP2())
+	go func() { server.Serve(grpcl) }()
```
Comment (heyitsanthony) [inline etcdmain/grpc_proxy.go]: `go func() { errc <- server.Serve(m.Match(cmux.HTTP2()) }()`?

File: etcdmain/grpc_proxy.go [hunk 3]
```diff
@@ -94,15 +109,34 @@ func startGRPCProxy(cmd *cobra.Command, args []string) {
...
 
-	server.Serve(l)
+	httpmux := http.NewServeMux()
+	httpmux.HandleFunc("/", http.NotFound)
+	httpmux.Handle("/metrics", prometheus.Handler())
+	srvhttp := &http.Server{
+		Handler: httpmux,
+	}
+	httpl := m.Match(cmux.HTTP1())
+	go func() { srvhttp.Serve(httpl) }()
```
Comment (heyitsanthony) [inline etcdmain/grpc_proxy.go]: `go func() { errc <- srvhttp.Serve(m.Match(cmux.HTTP1()) }()`?

File: etcdmain/grpc_proxy.go [hunk 4]
```diff
@@ -94,15 +109,34 @@ func startGRPCProxy(cmd *cobra.Command, args []string) {
...
+
+	if err := m.Serve(); err != nil {
```
Comment (heyitsanthony) [inline etcdmain/grpc_proxy.go]:
```go
go func() { errc <- m.Serve() }()
fmt.Fprintln(os.Stderr, <-errc)
os.Exit(1)
```
?

**Comment by heyitsanthony** [comment]
some nits about the error handling/log levels, but looks OK as-is

**Comment by gyuho** [comment]
All fixed. Will merge when all tests pass. Thanks!

File: etcdmain/grpc_proxy.go [hunk 5]
```diff
@@ -94,15 +109,36 @@ func startGRPCProxy(cmd *cobra.Command, args []string) {
...
+		grpc.StreamInterceptor(grpc_prometheus.StreamServerInterceptor),
+		grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor),
+	)
 	pb.RegisterKVServer(server, kvp)
 	pb.RegisterWatchServer(server, watchp)
 	pb.RegisterClusterServer(server, clusterp)
 	pb.RegisterLeaseServer(server, leasep)
 	pb.RegisterMaintenanceServer(server, mainp)
 	pb.RegisterAuthServer(server, authp)
+	grpc_prometheus.Register(server)
```
Comment (gyuho) [inline etcdmain/grpc_proxy.go]: Right, I will fix with tests in a separate PR. Thanks.

**Pull request merged**